File: src/test/java/io/github/bot/terminal/application/domains/db_connection/entity/DbConnectionTest.java
package io.github.bot.terminal.application.domains.db_connection.entity;

import io.github.bot.terminal.application.domains.db_connection.repository.DbConnectionRepository;
import io.github.bot.terminal.application.domains.db_connection.values.DbConnectionType;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.datasource.DriverManagerDataSource;

import javax.sql.DataSource;
import java.util.Optional;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class DbConnectionTest {

    @Mock
    private DbConnectionRepository repository;

    public String dbConnectionid = UUID.randomUUID().toString();
    public String name = UUID.randomUUID().toString();
    public String host = UUID.randomUUID().toString();
    public int port = 1233;
    public String user = UUID.randomUUID().toString();
    public String password = UUID.randomUUID().toString();
    public String dbName = UUID.randomUUID().toString();

    private DbConnection connection;

    @BeforeEach
    void init() {
        PostgresDbConnectionDetails details = new PostgresDbConnectionDetails();
        details.setId(dbConnectionid);
        details.setType(DbConnectionType.POSTGRES);
        details.setName(name);
        details.setHost(host);
        details.setPort(port);
        details.setUser(user);
        details.setPassword(password);
        details.setDbName(dbName);

        connection = spy(new PostgresDbConnection(details, repository));
    }

    @Test
    public void check() {
        JdbcTemplate templateMock = Mockito.mock(JdbcTemplate.class);
        doAnswer(invocation -> templateMock)
                .when(connection).createJdbcTemplate();

        connection.check();

        verify(templateMock, times(1)).execute(eq(connection.getCheckSelect()));
    }

    @Test
    public void createJdbcTemplate() {
        DataSource dataSource = Mockito.mock(DataSource.class);
        doAnswer(invocation -> Optional.of(dataSource))
                .when(connection).getDataSource();


        JdbcTemplate template = connection.createJdbcTemplate();
        assertEquals(dataSource, template.getDataSource());

    }

    @Test
    public void getDataSource() {
        String mockClassName = "java.lang.Class";
        connection.getDataSource();

        when(connection.getClassName()).thenReturn(mockClassName);
        DriverManagerDataSource dataSource = (DriverManagerDataSource)connection.getDataSource().get();

        assertEquals(connection.getUrl(), dataSource.getUrl() );
        assertEquals(connection.getDetails().getUser(), dataSource.getUsername() );
        assertEquals(connection.getDetails().getPassword(), dataSource.getPassword() );
        assertEquals(connection.getDetails().getPassword(), dataSource.getPassword() );
    }
}

File: src/test/kotlin/io/github/bot/terminal/application/domains/db_connection/entity/DbConnectionTest.kt
/*
package io.github.bot.terminal.application.domains.db_connection.entity

import io.github.bot.terminal.application.domains.db_connection.repository.DbConnectionRepository
import io.github.bot.terminal.application.domains.db_connection.values.DbConnectionType
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.ArgumentMatchers
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.invocation.InvocationOnMock
import org.mockito.junit.jupiter.MockitoExtension
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.jdbc.datasource.DriverManagerDataSource
import java.util.*
import javax.sql.DataSource

@ExtendWith(MockitoExtension::class)
internal class DbConnectionTest {
    @Mock
    private val repository: DbConnectionRepository? = null
    var dbConnectionid = UUID.randomUUID().toString()
    var name = UUID.randomUUID().toString()
    var host = UUID.randomUUID().toString()
    var port = 1233
    var user = UUID.randomUUID().toString()
    var password = UUID.randomUUID().toString()
    var dbName = UUID.randomUUID().toString()
    private var connection: DbConnection<*>? = null

    @BeforeEach
    fun init() {
        val details = PostgresDbConnectionDetails()
        details.setId(dbConnectionid)
        details.type = DbConnectionType.POSTGRES
        details.name = name
        details.host = host
        details.port = port
        details.user = user
        details.password = password
        details.dbName = dbName
        connection = Mockito.spy(PostgresDbConnection(details, repository!!))
    }

    @Test
    fun check() {
        val templateMock = Mockito.mock(JdbcTemplate::class.java)
        Mockito.doAnswer { invocation: InvocationOnMock? -> templateMock }
                .`when`(connection)!!.createJdbcTemplate()
        connection!!.check()
        Mockito.verify(templateMock, Mockito.times(1)).execute(ArgumentMatchers.eq(connection!!.checkSelect))
    }

    @Test
    fun createJdbcTemplate() {
        val dataSource = Mockito.mock(DataSource::class.java)
        Mockito.doAnswer { invocation: InvocationOnMock? -> Optional.of(dataSource) }
                .`when`(connection)!!.dataSource
        val template = connection!!.createJdbcTemplate()
        Assertions.assertEquals(dataSource, template.dataSource)
    }

    @get:Test
    val dataSource: Unit
        get() {
            val mockClassName = "java.lang.Class"
            connection!!.dataSource
            Mockito.`when`(connection!!.className).thenReturn(mockClassName)
            val dataSource = connection!!.dataSource.get() as DriverManagerDataSource
            Assertions.assertEquals(connection!!.url, dataSource.url)
            Assertions.assertEquals(connection!!.details.user, dataSource.username)
            Assertions.assertEquals(connection!!.details.password, dataSource.password)
            Assertions.assertEquals(connection!!.details.password, dataSource.password)
        }
}*/
